run tests with bundler since that is what our users run

grosser opened the pull request:
@nateberkopec would have caught https://github.com/puma/puma/issues/1308 early

does not change a lot since we previously ran the test with `bundle exec` and so already had all the load-paths etc setup with bundler

nateberkopec commented:
This seems to reliably cause the Ruby 2.1 integration tests to fail.

grosser commented:
test takes 43s locally ... that seems a bit crazy ... will take a look why ...

grosser commented:
works!

grosser commented:
wrong issue :D

grosser commented:
runs fast locally now .... still times out on travis :(

grosser commented:
reproduction steps with bundler 1.15.0+
```
gem uninstall puma -a
bundle exec puma test/rackup/hello.ru
ps -ef | grep puma
kill -SIGUSR2 <puma-pid>
```

grosser commented:
idk what's going on in travis, but it's cancelling all builds ...

grosser commented:
I guess it cancels all builds when 1 fails ... which won't work here since we have undocumented  know failures ...

grosser commented:
found the ruby 2.1 bug ... the Gemfile had no gemspec directive :D 
... having a look at jruby next

grosser commented:
green baby :D

grosser commented:
only issue left is now the handling of non-standard Gemfile paths ... I"ll address that in another PR

nateberkopec commented on lib/puma/launcher.rb:
```diff
@@ -163,7 +163,15 @@ def phased_restart
+        if defined?(Bundler)
+          env = Bundler::ORIGINAL_ENV
+          # add -rbundler/setup so we load from Gemfile when restarting
+          env["RUBYOPT"] = (env["RUBYOPT"].to_s.split(" ") << "-rbundler/setup").uniq.compact.join(" ")
```
Hmmm... are there scenarios where people's Rubyopt will get mangled by this?

nateberkopec replied:
This still makes me nervous. Why aren't we just appending this?

grosser commented:
Don't know how ... Maybe could duplicate the required but that would not
matter ... maybe if you load puma from a gem and then bundlerBut after ...
But would not be defined at this point then

On Jun 7, 2017 10:32 AM, "Nate Berkopec" <notifications@github.com> wrote:

> *@nateberkopec* commented on this pull request.
> ------------------------------
>
> In lib/puma/launcher.rb
> <https://github.com/puma/puma/pull/1317#discussion_r120693205>:
>
> > @@ -163,7 +163,15 @@ def phased_restart
>
>      # Run the server. This blocks until the server is stopped
>      def run
> -      previous_env = (defined?(Bundler) ? Bundler::ORIGINAL_ENV : ENV.to_h)
> +      previous_env =
> +        if defined?(Bundler)
> +          env = Bundler::ORIGINAL_ENV
> +          # add -rbundler/setup so we load from Gemfile when restarting
> +          env["RUBYOPT"] = (env["RUBYOPT"].to_s.split(" ") << "-rbundler/setup").uniq.compact.join(" ")
>
> Hmmm... are there scenarios where people's Rubyopt will get mangled by
> this?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/puma/puma/pull/1317#pullrequestreview-42677751>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ9kZJziZd4O_7nX4Ah-tMfSDkwVLks5sBt6ZgaJpZM4NvG7z>
> .
>

grosser commented:
@nateberkopec good to go ?

nateberkopec commented:
I'm fine with the test changes but mangling RUBYOPT just seems like a recipe for disaster.

grosser commented:
isn't just appending more dangerous ... could lead to duplicates
this way it's added unless it's there ... could rewrite to be more ifs and
less split/push/join style ... yeah the uniq could do something weird if
there are multiple -r in there ...

On Thu, Jun 8, 2017 at 6:48 PM, Nate Berkopec <notifications@github.com>
wrote:

> *@nateberkopec* commented on this pull request.
> ------------------------------
>
> In lib/puma/launcher.rb
> <https://github.com/puma/puma/pull/1317#discussion_r121034838>:
>
> > @@ -163,7 +163,15 @@ def phased_restart
>
>      # Run the server. This blocks until the server is stopped
>      def run
> -      previous_env = (defined?(Bundler) ? Bundler::ORIGINAL_ENV : ENV.to_h)
> +      previous_env =
> +        if defined?(Bundler)
> +          env = Bundler::ORIGINAL_ENV
> +          # add -rbundler/setup so we load from Gemfile when restarting
> +          env["RUBYOPT"] = (env["RUBYOPT"].to_s.split(" ") << "-rbundler/setup").uniq.compact.join(" ")
>
> This still makes me nervous. Why aren't we just appending this?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/puma/puma/pull/1317#discussion_r121034838>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZxrbqFtDj7wFyUTo9Zjm9y3yBMnyks5sCKSBgaJpZM4NvG7z>
> .
>

grosser commented:
updated to a safer approach

On Thu, Jun 8, 2017 at 7:48 PM, Michael Grosser <michael@grosser.it> wrote:

> isn't just appending more dangerous ... could lead to duplicates
> this way it's added unless it's there ... could rewrite to be more ifs and
> less split/push/join style ... yeah the uniq could do something weird if
> there are multiple -r in there ...
>
> On Thu, Jun 8, 2017 at 6:48 PM, Nate Berkopec <notifications@github.com>
> wrote:
>
>> *@nateberkopec* commented on this pull request.
>> ------------------------------
>>
>> In lib/puma/launcher.rb
>> <https://github.com/puma/puma/pull/1317#discussion_r121034838>:
>>
>> > @@ -163,7 +163,15 @@ def phased_restart
>>
>>      # Run the server. This blocks until the server is stopped
>>      def run
>> -      previous_env = (defined?(Bundler) ? Bundler::ORIGINAL_ENV : ENV.to_h)
>> +      previous_env =
>> +        if defined?(Bundler)
>> +          env = Bundler::ORIGINAL_ENV
>> +          # add -rbundler/setup so we load from Gemfile when restarting
>> +          env["RUBYOPT"] = (env["RUBYOPT"].to_s.split(" ") << "-rbundler/setup").uniq.compact.join(" ")
>>
>> This still makes me nervous. Why aren't we just appending this?
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/puma/puma/pull/1317#discussion_r121034838>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AAAsZxrbqFtDj7wFyUTo9Zjm9y3yBMnyks5sCKSBgaJpZM4NvG7z>
>> .
>>
>
>

nateberkopec commented:
Much better, thx.
